src/settings/global_user.rs

```diff
@@ -33,80 +125,122 @@ impl From<GlobalUser> for Credentials {
+    use tempfile::tempdir;
+
+    use crate::settings::environment::MockEnvironment;
 
     #[test]
     fn it_can_prioritize_token_input() {
         // Set all CF_API_TOKEN, CF_EMAIL, and CF_API_KEY.
         // This test evaluates whether the GlobalUser returned is
         // a GlobalUser::TokenAuth (expected behavior; token
         // should be prioritized over email + global API key pair.)
-        env::set_var("CF_API_TOKEN", "foo");
-        env::set_var("CF_EMAIL", "test@cloudflare.com");
-        env::set_var("CF_API_KEY", "bar");
+        let mut mock_env = MockEnvironment::default();
+        mock_env.set(CF_API_TOKEN, "foo");
+        mock_env.set(CF_EMAIL, "test@cloudflare.com");
+        mock_env.set(CF_API_KEY, "bar");
 
-        let user = get_global_config().unwrap();
+        let tmp_dir = tempdir().unwrap();
+        let config_dir = test_config_dir(&tmp_dir, None).unwrap();
+
+        let user = GlobalUser::build(mock_env, config_dir).unwrap();
         assert_eq!(
             user,
             GlobalUser::TokenAuth {
                 api_token: "foo".to_string()
             }
         );
     }
+
+    #[test]
+    fn it_can_prioritize_env_vars() {
+        let api_token = "thisisanapitoken";
+        let api_key = "reallylongglobalyapikey";
+        let email = "user@example.com";
+
+        let file_user = GlobalUser::TokenAuth {
+            api_token: api_token.to_string(),
+        };
+        let env_user = GlobalUser::GlobalKeyAuth {
+            api_key: api_key.to_string(),
+            email: email.to_string(),
+        };
+
+        let mut mock_env = MockEnvironment::default();
+        mock_env.set(CF_EMAIL, email);
+        mock_env.set(CF_API_KEY, api_key);
+
+        let tmp_dir = tempdir().unwrap();
+        let tmp_config_path = test_config_dir(&tmp_dir, Some(file_user)).unwrap();
+
+        let new_user = GlobalUser::build(mock_env, tmp_config_path).unwrap();
+
+        assert_eq!(new_user, env_user);
+    }
+
+    #[test]
+    fn it_falls_through_to_config_with_no_env_vars() {
+        let mock_env = MockEnvironment::default();
+
+        let user = GlobalUser::TokenAuth {
+            api_token: "thisisanapitoken".to_string(),
+        };
+
+        let tmp_dir = tempdir().unwrap();
+        let tmp_config_path = test_config_dir(&tmp_dir, Some(user.clone())).unwrap();
+
+        let new_user = GlobalUser::build(mock_env, tmp_config_path).unwrap();
+
+        assert_eq!(new_user, user);
+    }
+
+    #[test]
+    fn it_fails_if_global_auth_incomplete_in_env() {
+        let mut mock_env = MockEnvironment::default();
+
+        mock_env.set(CF_API_KEY, "apikey");
+
+        let tmp_dir = tempdir().unwrap();
+        let config_dir = test_config_dir(&tmp_dir, None).unwrap();
+
+        let new_user = GlobalUser::build(mock_env, config_dir);
+
+        match new_user {
+            Ok(_) => assert!(false),
+            Err(_) => assert!(true),
+        }
+    }
+
+    fn test_config_dir(
```

Comment:
i think from a readability perspective i would want this function to create the temp directory itself instead of having it passed in by every test, and i'd also want to call it something like `get_user_config_path` to match the name of the argument we're passing to `GlobalUser::build` which is `config_path`

Reply:
i would too! the problem is temp dirs die when they go out of scope